registry/storage/driver/hdfsweb/hdfs.go

```diff
@@ -0,0 +1,370 @@
+	}, nil
+}
+
+func (d *driver) Name() string {
+	return driverName
+}
+
+// GetContent retrieves the content stored at "path" as a []byte.
+func (d *driver) GetContent(ctx context.Context, subPath string) ([]byte, error) {
+	rc, err := d.ReadStream(ctx, subPath, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer rc.Close()
+
+	p, err := ioutil.ReadAll(rc)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+// PutContent stores the []byte content at a location designated by "path".
+func (d *driver) PutContent(ctx context.Context, subPath string, contents []byte) error {
+	hdfsFullPath := d.fullHdfsPath(subPath)
+	return d.fs.Create(
+		bytes.NewReader(contents),
+		hdfsFullPath)
+}
+
+// ReadStream retrieves an io.ReadCloser for the content stored at "path" with a
+// given byte offset.
+func (d *driver) ReadStream(ctx context.Context, subPath string, offset int64) (io.ReadCloser, error) {
+	hdfsFullPath := d.fullHdfsPath(subPath)
+
+	status, err := d.fs.GetFileStatus(hdfsFullPath)
+	if err != nil {
+		if reflect.TypeOf(err).Name() == "RemoteException" &&
+			err.(RemoteException).Exception == ExceptionFileNotFound {
+			return nil, storagedriver.PathNotFoundError{Path: subPath}
+		}
+		return nil, err
+	} else if status.Type != "FILE" {
+		return nil, fmt.Errorf("not file type")
+	} else if status.Length < offset {
+		return nil, storagedriver.InvalidOffsetError{Path: subPath, Offset: offset}
+	}
+
+	if status.Length == offset {
+		r := bytes.NewReader([]byte{})
+		rc := ioutil.NopCloser(r)
+		return rc, nil
+	}
+
+	reader, err := d.fs.Open(hdfsFullPath, offset, (status.Length - offset))
+	if err != nil {
+		return nil, err
+	}
+
+	return reader, nil
+}
+
+// WriteStream stores the contents of the provided io.Reader at a location
+// designated by the given path.
+func (d *driver) WriteStream(ctx context.Context, subPath string, offset int64, reader io.Reader) (int64, error) {
+	var (
+		buf          = bytes.NewBuffer(make([]byte, 0))
+		hdfsFullPath = d.fullHdfsPath(subPath)
+	)
+
+	if offset > 0 {
+		if status, err := d.fs.GetFileStatus(hdfsFullPath); err == nil && status.Type == "FILE" {
+			rc, err := d.ReadStream(ctx, subPath, 0)
+			if err != nil {
+				return 0, fmt.Errorf("Writing Failed %s", err)
+			}
+			if length, err := buf.ReadFrom(rc); err != nil {
+				return 0, err
+			} else if offset > length {
+				if _, err := buf.ReadFrom(bytes.NewReader((make([]byte, offset-length)))); err != nil {
+					return 0, err
+				}
+			} else if offset < length {
+				buf.Truncate(int(offset))
+			}
+		} else if status.Type != "FILE" ||
+			(reflect.TypeOf(err).Name() == "RemoteException" &&
+				err.(RemoteException).Exception == ExceptionFileNotFound) {
+			if _, err := buf.ReadFrom(bytes.NewReader((make([]byte, offset)))); err != nil {
+				return 0, err
+			}
+		} else {
+			return 0, err
+		}
+	}
+
+	sum, err := buf.ReadFrom(reader)
+	if err != nil {
+		return 0, err
+	}
+	if err := d.fs.Create(buf, hdfsFullPath); err != nil {
```

Comment:
You should be able to just pass in `reader` here instead of reading it all into `buf` first.